zerver/views/user_settings.py

```diff
@@ -94,9 +94,28 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
             # TODO: Make this into a user-facing error, not JSON
             raise UserDeactivatedError
 
+        # We re-validate all of the steps we did in
+        # json_change_settings, since the realm's settings, and list
+        # of active users, may have changed since then.
         if user_profile.realm.email_changes_disabled and not user_profile.is_realm_admin:
             raise JsonableError(_("Email address changes are disabled in this organization."))
 
+        error = validate_email_is_valid(
+            new_email,
+            get_realm_email_validator(user_profile.realm),
+        )
+        if error:
+            raise JsonableError(error)
+
+        try:
+            validate_email_not_already_in_realm(
+                user_profile.realm,
+                new_email,
+                verbose=False,
+            )
+        except ValidationError as e:
+            raise JsonableError(e.message)
```

Comment:
Rather than copying this, let's create a `validate_email_change_request(user_profile, new_email)` helper method and share that between the two code paths.